subsys/net/lib/ptp/transport.c

```diff
@@ -0,0 +1,354 @@
+
+		memcpy(&mreqn.imr_multiaddr, &mcast_addr, sizeof(struct in_addr));
+		mreqn.imr_ifindex = net_if_get_by_iface(port->iface);
+
+		zsock_setsockopt(port->socket[1], IPPROTO_IP,
+				 IP_ADD_MEMBERSHIP, &mreqn, sizeof(mreqn));
+	} else {
+		struct ipv6_mreq mreqn = {0};
+
+		memcpy(&mreqn.ipv6mr_multiaddr, &mcast_addr, sizeof(struct in6_addr));
+		mreqn.ipv6mr_ifindex = net_if_get_by_iface(port->iface);
+
+		zsock_setsockopt(port->socket[0], IPPROTO_IPV6,
+				 IPV6_ADD_MEMBERSHIP, &mreqn, sizeof(mreqn));
+	}
+
+	return 0;
+}
+
+static int transport_udp_ipv4_open(struct net_if *iface, uint16_t port)
+{
+	uint8_t tos;
+	socklen_t length;
+	int socket, ttl = 1;
+	struct sockaddr_in addr = {
+		.sin_family = AF_INET,
+		.sin_addr = INADDR_ANY_INIT,
+		.sin_port = htons(port),
+	};
+
+	socket = transport_socket_open(iface, (struct sockaddr *)&addr);
+
+	if (socket < 0) {
+		return -1;
+	}
+
+	if (zsock_setsockopt(socket, IPPROTO_IP, IP_MULTICAST_TTL, &ttl, sizeof(ttl))) {
+		LOG_ERR("Failed to set ip multicast ttl socket option");
+		goto error;
+	}
+
+	if (zsock_getsockopt(socket, IPPROTO_IP, IP_TOS, &tos, &length)) {
+		tos = 0;
+	}
+
+	tos &= ~0xFC;
+	tos = CONFIG_PTP_DSCP_VALUE << 2;
+	length = sizeof(tos);
+
+	if (zsock_setsockopt(socket, IPPROTO_IP, IP_TOS, &tos, length)) {
+		LOG_WRN("Failed to set DSCP priority");
+	}
+
+	return socket;
+error:
+	zsock_close(socket);
+	return -1;
+}
+
+static int transport_udp_ipv6_open(struct net_if *iface, uint16_t port)
+{
+	uint8_t tclass;
+	socklen_t length;
+	int socket, hops = 1, feature_on = 1;
+	struct sockaddr_in6 addr = {
+		.sin6_family = AF_INET6,
+		.sin6_addr = IN6ADDR_ANY_INIT,
+		.sin6_port = htons(port)
+	};
+
+	socket = transport_socket_open(iface, (struct sockaddr *)&addr);
+
+	if (socket < 0) {
+		return -1;
+	}
+
+	if (zsock_setsockopt(socket,
+			     IPPROTO_IPV6,
+			     IPV6_RECVPKTINFO,
+			     &feature_on,
+			     sizeof(feature_on))) {
+		LOG_ERR("Failed to set IPV6_RECVPKTINFO");
+		goto error;
+	}
+
+	if (zsock_setsockopt(socket, IPPROTO_IPV6, IPV6_MULTICAST_HOPS, &hops, sizeof(hops))) {
+		LOG_ERR("Failed to set ip multicast hops socket option");
+		goto error;
+	}
+
+	if (zsock_getsockopt(socket, IPPROTO_IPV6, IPV6_TCLASS, &tclass, &length)) {
+		tclass = 0;
+	}
+
+	tclass &= ~0xFC;
+	tclass = CONFIG_PTP_DSCP_VALUE << 2;
+	length = sizeof(tclass);
+
+	if (zsock_setsockopt(socket, IPPROTO_IPV6, IPV6_TCLASS, &tclass, length)) {
+		LOG_WRN("Failed to set priority");
+	}
+
+	return socket;
+error:
+	zsock_close(socket);
+	return -1;
+}
+
+static int transport_send(int socket, int port, void *buf, int length, struct sockaddr *addr)
+{
+	struct sockaddr m_addr;
+	socklen_t addrlen;
+	int cnt;
+
+	if (!addr) {
+		if (IS_ENABLED(CONFIG_PTP_UDP_IPv4_PROTOCOL)) {
+			m_addr.sa_family = AF_INET;
+			net_sin(&m_addr)->sin_port = htons(port);
+			net_sin(&m_addr)->sin_addr.s_addr = mcast_addr.s_addr;
+
+		} else if (IS_ENABLED(CONFIG_PTP_UDP_IPv6_PROTOCOL)) {
```

Comment:
I think this needs to be runtime configurable, so we should allow for example one interface to have IPv4, one IPv6 etc.

Reply:
agree, in the future I could implement something that would allow to specify protocol per port but for now I would like to proceed with one protocol for all ports.